### PR TITLE
HPCC-27961 Fix issue of allocating metric objects in DLLs that may be…

### PR DESCRIPTION
**Pull request by kenrowland**
… unloaded

Removed use of make_shared to separate object and control memory allocations

Signed-Off-By: Ken Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
@kenrowland looks good.
Please can you add a comment to make it clear why std::shared_ptr() is being used instead of make_shared - otherwise someone may come along and "optimize" it later.  Please squash the comment in, so I can merge once it is there.

**Comment by kenrowland** [comment]
@ghalliday Please see added comments surrounding the non use of make_shared. Also, I'm not getting notifications form Git, so it took me a while to realize you had reviewed. Hopefully I've solved that problem.

If the comments are OK, I'll rebase on master, squish, and request a merge

File: system/jlib/jmetrics.cpp [hunk 1]
```diff
@@ -304,3 +304,29 @@ void PeriodicMetricSink::doStopCollecting()
...
+    return createMetricAndAddToManager<GaugeMetric>(name, desc, units, metaData);
+}
+
+
+std::shared_ptr<GaugeMetricFromCounters> hpccMetrics::registerGaugeFromCountersMetric(const char *name, const char* desc, StatisticMeasure units,
+                                                                         const std::shared_ptr<CounterMetric> &pBeginCounter, const std::shared_ptr<CounterMetric> &pEndCounter,
+                                                                         const MetricMetaData &metaData)
+{
+    //
+    // std::make_shared avoided to ensure separate memory allocations for the object and control structures. This ensures the control structure is not
```
Comment (ghalliday) [inline system/jlib/jmetrics.cpp:326]: picky:  should be that it "ensures the vmt for the control structure does not live in the data belonging to the unloaded shared object."

**Review by ghalliday — COMMENTED** [review]
@kenrowland Although picky it is worth clarifying the comment.  (As it stands it is incorrect and potentially confusing.)

**Comment by kenrowland** [comment]
@ghalliday Updated the comment with information about the vmt. Please take a look.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**